src/vulkan-renderer/world/collision.cpp

```diff
@@ -0,0 +1,143 @@
+#include "inexor/vulkan-renderer/world/collision.hpp"
+
+#include <glm/gtx/intersect.hpp>
+#include <spdlog/spdlog.h>
+
+#include <algorithm>
+#include <cassert>
+#include <cmath>
+#include <unordered_map>
+
+namespace inexor::vulkan_renderer::world {
+
+CollisionData::CollisionData(std::shared_ptr<Cube> cube) : m_cube(cube) {}
+
+CollisionData::CollisionData() : m_cube(nullptr) {}
+
+void CollisionData::reset() {
+    m_cube = nullptr;
+}
+
+float CollisionData::squared_distance(glm::vec3 position) {
+    return std::pow(m_cube->center().x - position.x, 2) + std::pow(m_cube->center().y - position.y, 2) +
+           std::pow(m_cube->center().z - position.z, 2);
+}
+
+OctreeCollision::OctreeCollision(std::shared_ptr<Cube> cube) : m_cube(cube) {}
+
+bool OctreeCollision::ray_sphere_collision(const glm::vec3 position, const glm::vec3 direction) const {
+    auto distance = 0.0f;
+    const auto result =
+        glm::intersectRaySphere(position, direction, m_cube->center(), std::pow(m_cube->size() / 2, 2), distance);
+    return result;
+}
+
+bool OctreeCollision::ray_box_collision(const std::array<glm::vec3, 2> box_bounds, const glm::vec3 position,
+                                        const glm::vec3 direction) const {
+    glm::vec3 inverse_dir{1 / direction.x, 1 / direction.y, 1 / direction.z};
+    std::int32_t sign[3]{0, 0, 0};
+
+    sign[0] = (inverse_dir.x < 0);
+    sign[1] = (inverse_dir.y < 0);
+    sign[2] = (inverse_dir.z < 0);
+
+    float tmin, tmax, tymin, tymax, tzmin, tzmax;
+
+    tmin = (box_bounds[sign[0]].x - position.x) * inverse_dir.x;
+    tmax = (box_bounds[1 - sign[0]].x - position.x) * inverse_dir.x;
+    tymin = (box_bounds[sign[1]].y - position.y) * inverse_dir.y;
+    tymax = (box_bounds[1 - sign[1]].y - position.y) * inverse_dir.y;
+
+    if ((tmin > tymax) || (tymin > tmax)) {
+        return false;
+    }
+    if (tymin > tmin) {
+        tmin = tymin;
+    }
+    if (tymax < tmax) {
+        tmax = tymax;
+    }
+
+    tzmin = (box_bounds[sign[2]].z - position.z) * inverse_dir.z;
+    tzmax = (box_bounds[1 - sign[2]].z - position.z) * inverse_dir.z;
+
+    if ((tmin > tzmax) || (tzmin > tmax)) {
+        return false;
+    }
+    if (tzmin > tmin) {
+        tmin = tzmin;
+    }
+    if (tzmax < tmax) {
+        tmax = tzmax;
+    }
```

Comment:
(If not WIP): Why this code? It doesn't add any functionality, as `tmax` and `tmin`are method variables and the method returns true anyway.

Reply:
I removed this. It's unnecessary indeed.